diff --git a/Makefile b/Makefile
index d304602..0d62ff2 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 # Cross compile support - create a Makefile which defines these three variables and then includes this Makefile...
 CFLAGS  ?= -Wall -fPIC -O2 $(OPTS)
-LDFLAGS ?= -lasound -lpthread -lm -ldl -lrt
+EXTRALIBS ?= -lasound -lpthread -lm -ldl -lrt
 EXECUTABLE ?= squeezelite
 
 SOURCES = main.c slimproto.c utils.c output.c buffer.c stream.c decode.c process.c resample.c flac.c pcm.c mad.c vorbis.c faad.c mpg.c
@@ -11,7 +11,7 @@ OBJECTS = $(SOURCES:.c=.o)
 all: $(EXECUTABLE)
 
 $(EXECUTABLE): $(OBJECTS)
-	$(CC) $(OBJECTS) $(LDFLAGS) -o $@
+	$(CC) $(OBJECTS) $(LDFLAGS) $(EXTRALIBS) -o $@
 
 $(OBJECTS): $(DEPS)
 
